---
title: "Chemical analysis of microbiomes with chem16S and phyloseq"
author: "Jeffrey M. Dick"
output:
  html_vignette:
    mathjax: null
    toc: true
vignette: >
  %\VignetteIndexEntry{Chemical analysis of microbiomes with chem16S and phyloseq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: chem16S.bib
csl: elementa.csl
---

<script>
function ToggleDiv(ID) {
  var D = document.getElementById("D-" + ID);
  var B = document.getElementById("B-" + ID);
  if (D.style.display === "none") {
    // open the div and change button text
    D.style.display = "block";
    B.innerText = "Hide code";
  } else {
    // close the div and change button text
    D.style.display = "none";
    B.innerText = "Show code";
  }
}
</script>

```{r setup, include = FALSE}
options(width = 80)
## Use pngquant to optimize PNG images
library(knitr)
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL
```

```{r HTML, include = FALSE}
ZC <- "<i>Z</i><sub>C</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
```

This vignette demonstrates the integration of **chem16S** with **phyloseq** to calculate chemical metrics of community reference proteomes.

- *Chemical metrics* are molecular properties computed from elemental compositions -- inferred from amino acid compositions of proteins -- and include carbon oxidation state (`r ZC`) and stoichiometric hydration state (`r nH2O`), as described by @DYT20.
**chem16S** uses the `ZCAA()` and `H2OAA()` functions from the **canprot** package to calculate chemical metrics.
- *Community reference proteomes* are generated by combining lowest-level taxonomic assignments (from genus to phylum) with precomputed reference proteomes for taxa obtained from the RefSeq database, as described by @DT23.

## Load required packages

```{r load_packages}
library(chem16S)
library(phyloseq)
```

This vignette was compiled on `r Sys.Date()` with **chem16S** version `r sessionInfo()$otherPkgs$chem16S$Version` and **phyloseq** version `r sessionInfo()$otherPkgs$phyloseq$Version`.

## Data

The [mothur MiSeq SOP](https://mothur.org/wiki/miseq_sop/) has an example dataset with 16S rRNA sequences for mouse gut communities.
This example dataset has data collected at early () and late () times from one mouse.
This dataset is used in the [DADA2 Pipeline Tutorial](https://benjjneb.github.io/dada2/tutorial.html) to demonstrate construction of amplicon sequence variants (ASV), taxonomic classification of the ASVs, and analysis using the **phyloseq** package.
Here, we load the phyloseq object that was created on 2023-06-14 following the steps in the DADA2 tutorial (version 1.16).
Note that taxonomy was assigned to genus level using a newer version of the Silva training set ([`silva_nr99_v138.1_train_set.fa.gz`](https://zenodo.org/record/4587954)) than that indicated in the DADA2 tutorial on this date (`silva_nr_v132_train_set.fa.gz`).

```{r load_ps_silva}
ps_silva_file <- "extdata/DADA2/dada2_tutorial-silva_nr99_v138.1-phyloseq.Rdata"
load(system.file(ps_silva_file, package = "chem16S"))
ps_silva
```

## Lowest-level taxa

The **chem16S** package contains the amino acid compositions of archaeal, bacterial, and viral taxa at levels from genus to phylum constructed from the RefSeq database; see the vignette entitled [Chemical metrics of reference proteomes for taxa](reference-proteomes.html) for more information.
There are denoted as *reference proteomes* for taxa.
Several functions are available to identify the lowest-level taxon for each classified read (or in this case, each ASV), combine the reference proteomes for all taxa to create *community reference proteomes*, and calculate chemical metrics from them.

First, let's make a table that contains the lowest-level taxon and abundance of each ASV.
Because **chem16S** was first developed to analyze the output from the RDP Classifier, this data frame has a similar format to the file created by using the `classify -h` command of the RDP Classifier followed by the `merge-count` command.
That is, the data frame has columns named `taxid`, `rank`, `lineage`, and `name`, followed by columns for all samples.

```{r ps_silva_taxacounts}
tc <- ps_taxacounts(ps_silva)
```

Note that ASVs with identical lowest-level taxonomic assignments are merged, this data frame has fewer rows than the number of ASVs.
The short names of the ASVs are stored in the `taxid` column, and the names of multiple ASVs with the same lowest-level taxonomic classification are separated by a semicolon.
The ASVs themselves (DNA sequences, not short names) are available in the phyloseq object as explained in the DADA2 tutorial, but are not used here.
Now we can look at the number of lowest-level classifications at each taxonomic level:

```{r ps_silva_levels}
table(tc$rank)
```

## From taxonomy to chemical metrics

## Chemical analysis of genus-level assignments
Let's analyze the `GlobalPatterns` dataset that is part of **phyloseq**.

```{r GlobalPatterns}
data(GlobalPatterns)
```

## References
