---
title: "It ain’t ordination: Plotting two chemical variables"
author: "Jeffrey M. Dick"
output:
  html_document:
    mathjax: null
    toc: true
    css: style.css
vignette: >
  %\VignetteIndexEntry{It ain’t ordination: Plotting two chemical variables}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: chem16S.bib
csl: elementa.csl
---

<script>
function ToggleDiv(ID) {
  var D = document.getElementById("D-" + ID);
  var B = document.getElementById("B-" + ID);
  if (D.style.display === "none") {
    // open the div and change button text
    D.style.display = "block";
    B.innerText = "Hide code";
  } else {
    // close the div and change button text
    D.style.display = "none";
    B.innerText = "Show code";
  }
}
</script>


```{r HTML, include = FALSE}
Zc <- "<i>Z</i><sub>C</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
```

```{r setup, echo = FALSE}
options(width = 80)
## Use pngquant to optimize PNG images
library(knitr)
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL
# To make warnings appear in text box 20230619
# https://selbydavid.com/2017/06/18/rmarkdown-alerts/
knitr::knit_hooks$set(
   error = function(x, options) {
     paste('\n\n<div class="alert alert-danger">',
           gsub('##', '\n', gsub('^##\ Error', '**Error:**', x)),
           '</div>', sep = '\n')
   },
   warning = function(x, options) {
     paste('\n\n<div class="alert alert-warning">',
           gsub('##', '\n', gsub('^##\ Warning:', '**Warning:**', x)),
           '</div>', sep = '\n')
   },
   message = function(x, options) {
     paste('\n\n<div class="alert alert-info">',
           gsub('##', '\n', x),
           '</div>', sep = '\n')
   }
)
# Don't evaluate chunks if phyloseq is not available 20230619
if(!requireNamespace("phyloseq", quietly = TRUE)) {
  knitr::opts_chunk$set(eval = FALSE)
  warning("The **phyloseq** package is not available, so this vignette shows only the code without the results.")
}
```

What's the difference between *chemical metrics* and *diversity metrics*?
The visualization of diversity metrics for communities conventionally takes the form of ordinations, which are a type of dimensionality reduction that associates trends in the data with one or more synthetic variables.
Ordinations have a major limitation: the synthetic variables obtained in one study cannot be compared with those of other studies.

Chemical metrics quantify the elemental variation between communities by using reference protein sequences for entire genomes of sequenced organisms, multiplied by taxonomic abundances from 16R rRNA data.
Instead of synthetic variable names like "PCoA1" or "NMDS1", chemical metrics have names like "oxidation state" and "hydration state".
Because chemical metrics are defined independently of the data, they can be compared across studies.
Just as important, they can be used to explore hypotheses about environmental influences on elemental composition at the community level.

In this vignette we'll analyze the GlobalPatterns dataset from phyloseq [derived from @CLW+11] to get a glimpse into chemical variation of community reference proteomes across environments.
Then, we'll explore the hypothesis that chemical features of communities are shaped by the environmental chemistr by analyzing the OTU table and environmental measuremens including redox potential for Japanese hot springs reported by @OFY+19.

## Load required packages

```{r load_packages, message = FALSE}
require(chem16S)
require(phyloseq)
require(ggplot2)
theme_set(theme_bw())

# For composing plots and making a common legend (plot_layout())
require(patchwork)

# For annotating plots with regression coefficients (stat_poly_line())
require(ggpmisc)
```

This vignette was compiled on `r Sys.Date()` with **chem16S** version `r sessionInfo()$otherPkgs$chem16S$Version` and **phyloseq** version `r sessionInfo()$otherPkgs$phyloseq$Version`.

## GlobalPatterns

We will use the GlobalPatterns dataset 'as-is', without the preprocessing described in phyloseq's [Ordination Plots](https://joey711.github.io/phyloseq/plot_ordination-examples) tutorial.
There, less-abundant OTUs and phyla were removed in order to show high-level trends and shorten computing time.
One step we do borrow from that tutorial is the addition of a categorical variable that identifies whether the samples are human-associated:

```{r load_GlobalPatterns}
data(GlobalPatterns)
human = get_variable(GlobalPatterns, "SampleType") %in% c("Feces", "Mock", "Skin", "Tongue")
sample_data(GlobalPatterns)$human <- factor(human)
```

```{r plot_metrics2.GlobalPatterns, fig.width = 6, fig.height = 4, fig.align = "center"}
p2 <- plot_ps_metrics2(GlobalPatterns, color = "SampleType", shape = "human")
p2 + geom_polygon(aes(fill = SampleType)) + geom_point(size = 4)
```

At the extremes of carbon oxidation state (`r Zc`), soil communities are the most oxidized and skin and tongue communities are the most reduced.
At the extremes of stoichiometric hydration state (`r nH2O`), skin communities are the most hydrated and some fecal communities are the least hydrated.
In more detail, there are environmental microbiomes that show similar ranges of chemical metrics (e.g., Freshwater (creek) and Sediment (estuary)) and others that are different.
Freshwater (identified as "lake" in @CLW+11) has lower `r Zc` than Freshwater (creek), and some ocean samples have lower `r nH2O` than either freshwater group.
To a geochemical biologist, these patterns might suggest an influence of greater oxygenation in flowing water compared to lakes, or dehydration in communities adapted to life in salty water compared to freshwater.

## Humboldt Sulfuretum (Ocean Sediment)

@FEN+22 reported 16S rRNA sequences for sediment samples from the oxygen minimum zone of the Pacific Ocean along the coast of Chile, known as the Humboldt Sulfuretum.
The sample data included dissolved oxygen, redox potential in sediment and overlying water, and organic matter (OM) content.
This is a useful dataset for exploring the hypothesis that `r Zc` of proteins is shaped by redox conditions.
Here we read the phyloseq object created by using DADA2 to identify amplicon sequence variants (ASVs) in this dataset and classify them using the GTDB training set.
A sample taken from 50 m depth at the Valparaiso location on 2012-05-12 is available in the Sequence Read Archive (SRA) but was not included in the analysis described by @FEN+22.
The taxonomic composition of this sample is highly different from the all the others (see the ordination plots in the `extdata` directory where the `ps_FEN+22.rds` file is located), so we exclude it to avoid anomalous results.
```{r read_Humboldt}
psfile <- system.file("extdata/DADA2/FEN+22/ps_FEN+22.rds", package = "chem16S")
ps <- readRDS(psfile)
ps <- prune_samples(sample_names(ps) != "SRR1346095", ps)
```

Then, we plot `r Zc` and `r nH2O` for the community reference proteomes using different colors for sample locations.
```{r plot_metrics2.Humboldt, fig.width = 6, fig.height = 4, fig.align = "center"}
plot_ps_metrics2(ps, color = "Location") +
  geom_polygon(aes(fill = Location)) + geom_point(size = 4)
```

It appears that community reference proteomes for Iquique are, for the most part, more reduced (i.e., have lower `r Zc`) than those for Concepcion and Valparaiso.
Communities at Valparaiso are also characterized by higher `r nH2O`, suggesting the influence of a hydrating factor.

We are now equipped to explore hypotheses about physicochemical controls on elemental composition of community reference proteomes.
The next two plots show how sediment redox potential and concentration of organic matter vary among samples.
The rationale for choosing these environmental measurements is described below.

```{r check_patchwork, echo = FALSE}
# Don't evaluate remaining chunks if patchwork is not available 20230627
if(!requireNamespace("patchwork", quietly = TRUE)) {
  knitr::opts_chunk$set(eval = FALSE)
  warning("The **patchwork** package is not available, so the remaining plots are not shown.")
}
```

```{r plot_metrics2.Humboldt_redox_OM, fig.width = 6, fig.height = 5, fig.align = "center"}
p2 <- plot_ps_metrics2(ps, color = "Sediment_redox") + geom_point(size = 4)
p3 <- plot_ps_metrics2(ps, color = "Organic_matter") + geom_point(size = 4)
p2 / p3
```

Thermodynamic considerations predict a positive correlation between redox and `r Zc` [@DM23].
We could also predict that salinity may have a dehydrating effect that lowers `r nH2O` [@DYT20].
However, these samples have no documented salinity gradient; the only available salinity measurements for two samples indicate little difference from overlying seawater.
Previous observations that protein expression is shifted toward lower `r nH2O` under hyperglycemic (high-glucose) conditions [@DYT20] suggests another hypothesis: a higher content of organic matter may be a proxy for dehydrating conditions.

We can use correlations between two environmental variables (redox potential or OM) and two chemical metrics for communities (`r Zc` or `r nH2O`) in order to test these hypotheses.
To easily make the plots, let's make a data frame with the sample data and chemical metrics.
```{r sample.data.and.chemical.metrics.for.communities}
sample.data.and.chemical.metrics.for.communities <- cbind(sample_data(ps), ps_metrics(ps))
```

Now we write a function to create a scatter plot for two variables and add a regression line.
We use this function to make a plot for each combination of environmental variables and chemical metrics.

```{r check_ggpmisc, echo = FALSE}
# Don't evaluate remaining chunks if ggpmisc is not available 20230627
if(!requireNamespace("ggpmisc", quietly = TRUE)) {
  knitr::opts_chunk$set(eval = FALSE)
  warning("The **ggpmisc** package is not available, so the remaining plots are not shown.")
}
```

<button id="B-scatter_plot" onclick="ToggleDiv('scatter_plot')">Show code</button>
<div id="D-scatter_plot" style="display: none">
```{r scatter_plot, eval = FALSE}
# Defuse (enquo) and Inject (!!) from https://www.tidyverse.org/blog/2018/07/ggplot2-tidy-evaluation/
# Regression line and equation from https://stackoverflow.com/questions/7549694/add-regression-line-equation-and-r2-on-graph
scatter_plot <- function(data = sample.data.and.chemical.metrics.for.communities, x, y) {
  x <- enquo(x)
  y <- enquo(y)
  ggplot(data, aes(x = !!x, y = !!y, color = .data[["Location"]])) +
    # Override aes to plot one regression line for samples from all locations
    stat_poly_line(aes(x = !!x, y = !!y), inherit.aes = FALSE) +
    stat_poly_eq(aes(x = !!x, y = !!y), inherit.aes = FALSE) +
    geom_point()
}

sp1 <- scatter_plot(x = Sediment_redox, y = Zc)
sp2 <- scatter_plot(x = Sediment_redox, y = nH2O)
sp3 <- scatter_plot(x = Organic_matter, y = Zc)
sp4 <- scatter_plot(x = Organic_matter, y = nH2O)
sp1 + sp2 + sp3 + sp4 + plot_layout(guides = "collect")
```
</div>
```{r scatter_plot, echo = FALSE, fig.width = 7, fig.height = 6, fig.align = "center"}
```

We see find that carbon oxidation state is positively correlated with redox potential, and stoichiometric hydration state is negatively correlated with organic matter content.
Taken alone, each of these correlations supports our initial hypotheses.
However, `r Zc` is also negatively correlated with OM content, and `r nH2O` is positively correlated with redox potential.
The reason is that the environmental variables strongly covary; this is a common feature of environmental data.

The covariation of environmental variables makes it difficult to identify primary factors that drive the observed differences between communities.
However, recognizing the chemical nature of these variables provides new insight into processes that may shape the adaptation of microbial genomes.
It is notable that samples with higher OM content have lower redox potential, and vice versa.
This makes sense if greater availability of organic compounds drives respiration and ensuing depletion of oxygen.
This interaction among environmental variables provides a plausible explanation for the positive association between `r Zc` and `r nH2O` (see first plot above).
Our initial hypotheses about influences of redox potential and organic matter taken alone could not explain that association, so this is a new hypothesis that integrates more physicochemical knowledge.
To test this new hypothesis, future analysis should be done with datasets where either redox potential or OM content are controlled (i.e., exhibit no variation) in order to identify the effect of each variable alone.


## Take-home messages

* Unlike synthetic variables in ordination, chemical metrics are defined independently of the data and can be compared across datasets.
* Trends between `r Zc` and `r nH2O` suggest the influence not of single environmental variables but of interactions between variables, such as lower redox potential associated with greater organic matter content in sediments.
* Variation in `r nH2O` may reflect the influence of various factors such as salinity or organic matter content.
  In the GlobalPatterns dataset, some fecal samples have extremely low `r nH2O`, which may be a signature of adaptation to dehydrating conditions.
* Higher `r Zc` is associated with greater redox potential in the Humboldt Sulfuretum dataset.

## References
