---
title: "Chemical metrics of reference proteomes for taxa"
author: "Jeffrey M. Dick"
output:
  html_vignette:
    mathjax: null
    toc: true
vignette: >
  %\VignetteIndexEntry{Chemical metrics of reference proteomes for taxa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: chem16S.bib
csl: elementa.csl
link-citations: true
---

```{r setup, include = FALSE}
options(width = 80)
## Use pngquant to optimize PNG images
library(knitr)
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL
```

```{r HTML, include = FALSE}
Zc <- "<i>Z</i><sub>C</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
```

Here we visualize chemical metrics for genera aggregated to higher levels, using precomputed reference proteomes for taxa in the **chem16S** package.

> *Chemical metrics* are molecular properties computed from elemental compositions (derived from amino acid compositions) of proteins and include carbon oxidation state (`r Zc`) and stoichiometric hydration state (`r nH2O`), as described by @DYT20.

This vignette uses the RefSeq reference database for reference proteomes.
The GTDB reference database is also available in **chem16S** (and is the default for functions in the package), but doesn't have viral reference proteomes, which are visualized below.

## Required packages

```{r load_packages, message = FALSE}
library(chem16S)
```

This vignette was compiled on `r Sys.Date()` with **chem16S** version `r sessionInfo()$otherPkgs$chem16S$Version`.

## Reference proteomes for taxa

* Read the acid compositions of reference proteomes.
* Show the counts of taxa at each level (taxonomic rank).
```{r taxon_AA}
taxon_AA <- read.csv(system.file("extdata/RefSeq/taxon_AA.csv.xz", package = "chem16S"))
ranks <- taxon_AA$protein
table(ranks)[unique(ranks)]
```

* Calculate `r Zc`.
* Make boxplots for the taxa at each rank.
```{r Zc_boxplot, fig.width = 5, fig.height = 5, fig.align = "center", pngquant = pngquant}
taxon_Zc <- calc_metrics(taxon_AA, "Zc")[, 1]
Zc_list <- sapply( unique(ranks), function(rank) taxon_Zc[ranks == rank] )
par(mar = c(4, 7, 1, 1))
boxplot(Zc_list, horizontal = TRUE, las = 1, xlab = chemlab("Zc"))
```

## `r Zc` for genera in selected phyla and classes

* Read the list of taxonomic names.
* Define a function to get the names of unique genera in a phylum.
* Define a function to get `r Zc` for named genera.
* Calculate mean `r Zc` for genera in selected phyla.
```{r phylum_to_genus}
taxnames <- read.csv(system.file("extdata/RefSeq/taxon_names.csv.xz", package = "chem16S"))
phylum_to_genus <- function(phylum) na.omit(unique(taxnames$genus[taxnames$phylum == phylum]))
get_Zc <- function(genera) na.omit(taxon_Zc[match(genera, taxon_AA$organism)])
sapply(sapply(sapply(c("Crenarchaeota", "Euryarchaeota"), phylum_to_genus), get_Zc), mean)
```

Within the Euryarchaeota, there are classes with extremely high and low `r Zc` [see below and @DT23]. Let's look at a couple of them:
```{r class_to_genus}
class_to_genus <- function(class) na.omit(unique(taxnames$genus[taxnames$class == class]))
sapply(sapply(sapply(c("Methanococci", "Halobacteria"), class_to_genus), get_Zc), mean)
```

## Prokaryotic genera, aggregated to phylum

Read full list of taxonomic names; remove viruses; prune to keep unique genera; count number of genera in each phylum; calculate `r Zc` for genera in 20 most highly represented phyla of Bacteria and Archaea; order according to mean `r Zc`; make boxplots:
```{r phylum_Zc, fig.width = 7, fig.height = 6, fig.align = "center", pngquant = pngquant}
taxnames2 <- taxnames[taxnames$superkingdom != "Viruses", ]
taxnames3 <- taxnames2[!duplicated(taxnames2$genus), ]
(top20_phyla <- head(sort(table(taxnames3$phylum), TRUE), 20))
Zc_list <- sapply(sapply(names(top20_phyla), phylum_to_genus), get_Zc)
order_Zc <- order(sapply(Zc_list, mean))
Zc_list <- Zc_list[order_Zc]
par(mar = c(4, 13, 1, 1))
boxplot(Zc_list, horizontal = TRUE, las = 1, xlab = chemlab("Zc"))
```

## Prokaryotic genera, aggregated to class

Notice the large range for Euryarchaota and Protobacteria.
Let's take a closer look at the classes within each phylum.

* Read full list of taxonomic names.
* Remove viruses.
* Prune to keep unique genera.
* Count number of genera in each class.
* Calculate `r Zc` for genera in 10 most highly represented classes of Bacteria and Archaea.
* Order according to mean `r Zc`.
* Make boxplots.
```{r class_Zc, fig.width = 10, fig.height = 5, fig.align = "center", pngquant = pngquant}
par(mfrow = c(1, 2))
for(phylum in c("Euryarchaeota", "Proteobacteria")) {
  taxnames4 <- taxnames3[taxnames3$phylum == phylum, ]
  (top10_classes <- head(sort(table(taxnames4$class), TRUE), 10))
  Zc_list <- sapply(sapply(names(top10_classes), class_to_genus), get_Zc)
  order_Zc <- order(sapply(Zc_list, mean))
  Zc_list <- Zc_list[order_Zc]
  par(mar = c(4, 10, 1, 1))
  boxplot(Zc_list, horizontal = TRUE, las = 1, xlab = chemlab("Zc"))
}
```

*See take-home message #1.*

## Stoichiometric hydration state (`r nH2O`)

```{r class_nH2O, fig.width = 10, fig.height = 5, fig.align = "center", pngquant = pngquant}
taxon_nH2O <- calc_metrics(taxon_AA, "nH2O")[, 1]
get_nH2O <- function(genera) na.omit(taxon_nH2O[match(genera, taxon_AA$organism)])

par(mfrow = c(1, 2))
for(phylum in c("Euryarchaeota", "Proteobacteria")) {
  taxnames4 <- taxnames3[taxnames3$phylum == phylum, ]
  (top10_classes <- head(sort(table(taxnames4$class), TRUE), 10))
  nH2O_list <- sapply(sapply(names(top10_classes), class_to_genus), get_nH2O)
  order_nH2O <- order(sapply(nH2O_list, mean))
  nH2O_list <- nH2O_list[order_nH2O]
  par(mar = c(4, 10, 1, 1))
  boxplot(nH2O_list, horizontal = TRUE, las = 1, xlab = chemlab("nH2O"))
}
```

*See take-home message #2.*

## Viruses and prokaryotes

* Plot `r Zc` and `r nH2O` for viral and prokaryotic phyla with the most genus-level representatives.
* Label Bacteria with the lowest `r nH2O`.
```{r viruses, fig.width = 7, fig.height = 5, fig.align = "center", pngquant = pngquant}
(top50_phyla <- head(sort(table(taxnames$phylum), TRUE), 50))
Zc_mean <- sapply(sapply(sapply(names(top50_phyla), phylum_to_genus), get_Zc), mean)
nH2O_mean <- sapply(sapply(sapply(names(top50_phyla), phylum_to_genus), get_nH2O), mean)
domain <- taxnames$superkingdom[match(names(top50_phyla), taxnames$phylum)]
pchs <- c(24, 21, 23)
pch <- sapply(domain, switch, Archaea = pchs[1], Bacteria = pchs[2], Viruses = pchs[3])
bgs <- topo.colors(3, alpha = 0.5)
bg <- sapply(domain, switch, Archaea = bgs[1], Bacteria = bgs[2], Viruses = bgs[3])
par(mar = c(4, 4, 1, 1))
plot(Zc_mean, nH2O_mean, xlab = chemlab("Zc"), ylab = chemlab("nH2O"), pch = pch, bg = bg)
ilow <- nH2O_mean < -0.77 & domain == "Bacteria"
xadj <- c(-0.9, -0.8, 0.8, 1, -0.8)
yadj <- c(0, 1, 1, -1, -1)
text(Zc_mean[ilow] + 0.02 * xadj, nH2O_mean[ilow] + 0.005 * yadj, names(top50_phyla[ilow]), cex = 0.9)
legend("bottomleft", c("Archaea", "Bacteria", "Viruses"), pch = pchs, pt.bg = bgs)
```

*See take-home message #3.*

## Take-home messages

1. Methanococci and Epsilonproteobacteria (now known as Campylobacterota) are prokaryotic classes whose genera have proteins with the lowest mean `r Zc` in their respective phyla.
   This may suggest adaptation to reducing environments such as submarine hot springs and anoxic zones of sediments.
2. Proteins in Gammaproteobacteria have lower `r nH2O` than Alpha- and Betaproteobacteria.
   This may suggest adaptation to lower water availability in certain habitats.
3. Viral proteins have lower `r nH2O` than most Bacteria and Archaea.
   This may suggest adaptation to lower water availability in the molecular environment of viruses.
   Notably, viruses without an envelope have lower water content than bacterial cells [@Mat75].

In summary, chemical metrics can reveal new insights into environmental factors that shape elemental composition of protein sequences.
This is a new method in comparative genomics that can help to understand geochemical effects on genome evolution, or "geochemical biology".

## References
